File: backend/app/schemas/ai_script_structure.py
"""
Schemas Pydantic para AIScriptStructure.

Representación intermedia editable de un JMX. NO se persiste en DB:
se deriva on-demand del current_jmx con el parser del Sprint 2.1.

Modelo basado en análisis real de sample_jmx/Ejercicio_Booking.jmx.

Sprint 2.3a: campo `is_dirty: bool = False` añadido a 24 modelos para
habilitar la estrategia edit-preserving del regenerador
(`structure_to_jmx.regenerate_jmx_from_structure`).
"""
from typing import List, Optional, Literal, Union, Dict, Any
from uuid import UUID, uuid4
from datetime import datetime

from pydantic import BaseModel, Field, ConfigDict


# ============================================================================
# Configuración base — habilita aliases con espacios (ej. "Start users count")
# ============================================================================

class JMXBase(BaseModel):
    model_config = ConfigDict(
        populate_by_name=True,         # acepta tanto field name como alias
        extra='allow',                  # passthrough de campos no contemplados
    )


# ============================================================================
# Test Plan (raíz)
# ============================================================================

class TestPlanModel(JMXBase):
    name: str = "Test Plan"
    functional_mode: bool = False
    serialize_threadgroups: bool = False
    tearDown_on_shutdown: bool = True
    comments: Optional[str] = None
    is_dirty: bool = False


# ============================================================================
# Variables top-level (Arguments / UserDefinedVariables)
# ============================================================================

class UserDefinedVariable(JMXBase):
    name: str
    value: str
    metadata: str = "="
    description: Optional[str] = None
    is_dirty: bool = False


# ============================================================================
# HTTP Defaults
# ============================================================================

class HttpDefaultsModel(JMXBase):
    domain: Optional[str] = None
    protocol: Optional[str] = None
    port: Optional[str] = None
    path: Optional[str] = None
    implementation: Optional[str] = None
    encoding: Optional[str] = None
    is_dirty: bool = False


# ============================================================================
# Cookie / Cache Manager
# ============================================================================

class CookieManagerModel(JMXBase):
    enabled: bool = True
    clear_each_iteration: bool = False
    policy: Optional[str] = None
    is_dirty: bool = False


class CacheManagerModel(JMXBase):
    enabled: bool = True
    clear_each_iteration: bool = False
    use_expires: bool = True
    is_dirty: bool = False


# ============================================================================
# CSV Data Sets
# ============================================================================

class CSVDataSetModel(JMXBase):
    id: UUID = Field(default_factory=uuid4)
    testname: str
    enabled: bool = True
    filename: str
    variable_names: List[str] = []
    delimiter: str = ","
    ignore_first_line: bool = False
    quoted_data: bool = False
    recycle: bool = True
    share_mode: Literal["shareMode.all", "shareMode.group", "shareMode.thread"] = "shareMode.all"
    stop_thread: bool = False
    file_encoding: Optional[str] = None
    is_dirty: bool = False


# ============================================================================
# Stepping Thread Group config (subobjeto opcional)
# Nombres con espacios via alias — preserva fidelidad del XML kg.apc
# ============================================================================

class SteppingConfig(JMXBase):
    initial_delay: int = Field(default=0, alias="Threads initial delay")
    start_users_count: int = Field(default=1, alias="Start users count")
    start_users_count_burst: int = Field(default=0, alias="Start users count burst")
    start_users_period: int = Field(default=30, alias="Start users period")
    stop_users_count: int = Field(default=1, alias="Stop users count")
    stop_users_period: int = Field(default=5, alias="Stop users period")
    ramp_up: int = Field(default=5, alias="rampUp")
    flight_time: int = Field(default=60, alias="flighttime")


# ============================================================================
# Sampler children — modelos discriminados por 'type'
# ============================================================================

class HeaderModel(JMXBase):
    name: str
    value: str


class HeaderManagerModel(JMXBase):
    id: UUID = Field(default_factory=uuid4)
    enabled: bool = True
    headers: List[HeaderModel] = []
    is_dirty: bool = False


class ResponseAssertionModel(JMXBase):
    id: UUID = Field(default_factory=uuid4)
    enabled: bool = True
    name: str = "Response Assertion"
    test_field: str = "Assertion.response_data"  # response_data | response_code | response_headers | response_message
    test_type: int = 2                            # bitmask: 1=matches, 2=contains, 4=not, 8=equals, 16=substring
    test_strings: List[str] = []
    custom_message: Optional[str] = None
    assume_success: bool = False
    # Flags derivados (computed) — NO se serializan al JMX, son para UI:
    negate: bool = False                          # set true si bit 4 está activo en test_type
    pattern_match: Literal["contains", "matches", "equals", "substring"] = "contains"
    is_dirty: bool = False


class RegexExtractorModel(JMXBase):
    id: UUID = Field(default_factory=uuid4)
    enabled: bool = True
    name: str = "Regex Extractor"
    refname: str
    regex: str
    template: str = "$1$"
    match_number: str = "1"
    default: str = ""
    default_empty_value: bool = False
    use_headers: Literal["false", "true", "URL", "code", "message"] = "false"
    scope: Optional[Literal["all", "parent", "children"]] = None
    extract_from: Literal["body", "header", "url", "code", "message"] = "body"  # derivado de use_headers
    is_dirty: bool = False


class JsonExtractorModel(JMXBase):
    id: UUID = Field(default_factory=uuid4)
    enabled: bool = True
    name: str = "JSON Extractor"
    refname: str
    json_path: str
    match_number: str = "1"
    default: str = ""
    is_dirty: bool = False


class XPathExtractorModel(JMXBase):
    id: UUID = Field(default_factory=uuid4)
    enabled: bool = True
    name: str = "XPath Extractor"
    refname: str
    xpath: str
    default: str = ""
    is_dirty: bool = False


class BoundaryExtractorModel(JMXBase):
    id: UUID = Field(default_factory=uuid4)
    enabled: bool = True
    name: str = "Boundary Extractor"
    refname: str
    left_boundary: str
    right_boundary: str
    match_number: str = "1"
    default: str = ""
    is_dirty: bool = False


class ConstantTimerModel(JMXBase):
    id: UUID = Field(default_factory=uuid4)
    enabled: bool = True
    name: str = "Constant Timer"
    delay_ms: int = 0
    is_dirty: bool = False


class UniformRandomTimerModel(JMXBase):
    id: UUID = Field(default_factory=uuid4)
    enabled: bool = True
    name: str = "Uniform Random Timer"
    constant_delay_ms: int = 0
    random_delay_ms: int = 100
    is_dirty: bool = False


class GaussianRandomTimerModel(JMXBase):
    id: UUID = Field(default_factory=uuid4)
    enabled: bool = True
    name: str = "Gaussian Random Timer"
    constant_delay_ms: int = 0
    deviation_ms: int = 100
    is_dirty: bool = False


# ============================================================================
# HTTP Sampler body
# ============================================================================

class FormArgument(JMXBase):
    name: str
    value: str
    always_encode: bool = False
    use_equals: bool = True
    metadata: str = "="


class SamplerBody(JMXBase):
    mode: Literal["raw", "form", "none"] = "none"
    raw_text: Optional[str] = None
    form_args: List[FormArgument] = []
    body_type: Literal["json", "xml", "form", "raw", "auto"] = "auto"  # heurística según Content-Type


# ============================================================================
# Sampler child wrapper (discriminado por 'type')
# ============================================================================

SamplerChildType = Literal[
    "header_manager",
    "response_assertion",
    "regex_extractor",
    "json_extractor",
    "xpath_extractor",
    "boundary_extractor",
    "constant_timer",
    "uniform_random_timer",
    "gaussian_random_timer",
    "unsupported",
]


class UnsupportedElement(JMXBase):
    """Passthrough para elementos no soportados — preserva raw_xml + warning."""
    id: UUID = Field(default_factory=uuid4)
    type: Literal["unsupported"] = "unsupported"
    kind: str                                    # ej. "JSR223PreProcessor"
    name: Optional[str] = None
    reason: str                                  # ej. "PreProcessor JSR223 no editable visualmente"
    severity: Literal["warning", "info"] = "warning"
    raw_xml: str                                 # bloque XML original para round-trip
    parent_sampler_id: Optional[UUID] = None


class SamplerChild(JMXBase):
    """
    Wrapper discriminado por 'type'. El campo 'data' contiene el modelo concreto.
    Esto permite preservar el ORDEN de los children, que JMeter usa
    semánticamente (ej. extractor antes que assertion = el assertion
    puede usar la variable extraída).
    """
    type: SamplerChildType
    order: int                                   # posición en el hashTree
    data: Union[
        HeaderManagerModel,
        ResponseAssertionModel,
        RegexExtractorModel,
        JsonExtractorModel,
        XPathExtractorModel,
        BoundaryExtractorModel,
        ConstantTimerModel,
        UniformRandomTimerModel,
        GaussianRandomTimerModel,
        UnsupportedElement,
    ]


# ============================================================================
# HTTP Sampler
# ============================================================================

class HTTPSamplerModel(JMXBase):
    id: UUID = Field(default_factory=uuid4)
    type: Literal["sampler"] = "sampler"
    enabled: bool = True
    name: str
    method: str = "GET"                          # GET/POST/PUT/DELETE/PATCH/HEAD/OPTIONS
    protocol: Optional[str] = None
    domain: Optional[str] = None
    port: Optional[str] = None
    path: str = ""
    content_encoding: str = "UTF-8"
    follow_redirects: bool = True
    use_keepalive: bool = True
    auto_redirects: bool = False
    body: SamplerBody = Field(default_factory=SamplerBody)
    children: List[SamplerChild] = []            # orden preservado
    raw_xml: str = ""                            # backup completo del sampler para round-trip
    is_dirty: bool = False


# ============================================================================
# Controllers (Top 5 — GenericController, LoopController, IfController,
# WhileController, ThroughputController). ForEach queda v2.
# ============================================================================

ControllerType = Literal[
    "generic_controller",
    "loop_controller",
    "if_controller",
    "while_controller",
    "throughput_controller",
]


class GenericControllerModel(JMXBase):
    id: UUID = Field(default_factory=uuid4)
    type: Literal["controller"] = "controller"
    kind: Literal["generic_controller"] = "generic_controller"
    enabled: bool = True
    name: str
    children: List["TGChild"] = []
    is_dirty: bool = False


class LoopControllerModel(JMXBase):
    id: UUID = Field(default_factory=uuid4)
    type: Literal["controller"] = "controller"
    kind: Literal["loop_controller"] = "loop_controller"
    enabled: bool = True
    name: str
    loops: int = 1
    continue_forever: bool = False
    children: List["TGChild"] = []
    is_dirty: bool = False


class IfControllerModel(JMXBase):
    id: UUID = Field(default_factory=uuid4)
    type: Literal["controller"] = "controller"
    kind: Literal["if_controller"] = "if_controller"
    enabled: bool = True
    name: str
    condition: str
    use_expression: bool = True
    evaluate_all: bool = False
    children: List["TGChild"] = []
    is_dirty: bool = False


class WhileControllerModel(JMXBase):
    id: UUID = Field(default_factory=uuid4)
    type: Literal["controller"] = "controller"
    kind: Literal["while_controller"] = "while_controller"
    enabled: bool = True
    name: str
    condition: str
    children: List["TGChild"] = []
    is_dirty: bool = False


class ThroughputControllerModel(JMXBase):
    id: UUID = Field(default_factory=uuid4)
    type: Literal["controller"] = "controller"
    kind: Literal["throughput_controller"] = "throughput_controller"
    enabled: bool = True
    name: str
    style: int = 0                               # 0=percent, 1=total
    percent_throughput: str = "100.0"
    per_thread: bool = False
    children: List["TGChild"] = []
    is_dirty: bool = False


# Union de Controllers
ControllerUnion = Union[
    GenericControllerModel,
    LoopControllerModel,
    IfControllerModel,
    WhileControllerModel,
    ThroughputControllerModel,
]


# ============================================================================
# Thread Group child — discriminado por 'type'
# ============================================================================

class TGChild(JMXBase):
    """
    Hijo de Thread Group o Controller. Puede ser sampler, controller anidado,
    o elemento no soportado. Discriminado por 'type'.
    """
    type: Literal["sampler", "controller", "unsupported"]
    order: int
    sampler: Optional[HTTPSamplerModel] = None
    controller: Optional[ControllerUnion] = None
    unsupported: Optional[UnsupportedElement] = None


# Forward refs para self-referencing controllers
GenericControllerModel.model_rebuild()
LoopControllerModel.model_rebuild()
IfControllerModel.model_rebuild()
WhileControllerModel.model_rebuild()
ThroughputControllerModel.model_rebuild()


# ============================================================================
# Thread Group
# ============================================================================

class ThreadGroupModel(JMXBase):
    id: UUID = Field(default_factory=uuid4)
    kind: Literal["standard", "stepping", "concurrency", "ultimate"] = "standard"
    name: str = "Thread Group"
    enabled: bool = True                         # crítico: smoke vs carga se distingue por enabled
    comments: Optional[str] = None
    on_sample_error: Literal[
        "continue", "startnextloop", "stopthread", "stoptest", "stoptestnow"
    ] = "continue"
    num_threads: int = 1
    ramp_time: int = 1
    duration: Optional[int] = None
    delay: Optional[int] = None
    scheduler: bool = False
    loops: int = 1
    continue_forever: bool = False
    stepping: Optional[SteppingConfig] = None    # solo si kind == "stepping"
    children: List[TGChild] = []                 # samplers + controllers en orden
    raw_xml: str = ""                            # backup completo
    is_dirty: bool = False


# ============================================================================
# Config elements (top-level: CookieManager, CacheManager, AuthManager, etc.)
# ============================================================================

ConfigElementKind = Literal[
    "cookie_manager",
    "cache_manager",
    "auth_manager",
    "dns_cache_manager",
    "keystore_config",
    "other",
]


class ConfigElementModel(JMXBase):
    id: UUID = Field(default_factory=uuid4)
    kind: ConfigElementKind
    name: str
    enabled: bool = True
    properties: Dict[str, Any] = {}
    raw_xml: str = ""
    is_dirty: bool = False


# ============================================================================
# Listeners (incluye kg.apc family + estándar)
# ============================================================================

ListenerKind = Literal[
    "view_results_tree",
    "summary_report",
    "aggregate_report",
    "graph_results",
    "kg_apc_response_times_over_time",
    "kg_apc_response_codes_per_second",
    "kg_apc_transactions_per_second",
    "kg_apc_active_threads_over_time",
    "kg_apc_hits_per_second",
    "other",
]


class ListenerModel(JMXBase):
    id: UUID = Field(default_factory=uuid4)
    kind: ListenerKind
    guiclass: str                                # guiclass original ("ViewResultsFullVisualizer" etc.)
    name: str
    enabled: bool = True
    filename: Optional[str] = None
    raw_xml: str = ""                            # passthrough — save_config props internas
    is_dirty: bool = False


# ============================================================================
# Metadata derivada (calculada por el parser)
# ============================================================================

class StructureMetadata(JMXBase):
    jmx_version: str = "5.6.3"
    parsed_at: datetime = Field(default_factory=datetime.utcnow)
    referenced_variables: List[str] = []         # ${var} encontradas en cualquier sampler
    defined_variables: List[str] = []            # union de UDV + CSV columns + extractor refnames
    undefined_variables: List[str] = []          # referenced - defined → warning crítico
    has_unmapped: bool = False
    unmapped_count: int = 0
    parse_warnings: List[str] = []               # warnings no fatales del parser


# ============================================================================
# Modelo raíz: AIScriptStructure
# ============================================================================

class AIScriptStructure(JMXBase):
    """
    Representación intermedia editable de un JMX completo.
    Se deriva on-demand desde current_jmx (no se persiste en DB).
    """
    test_plan: TestPlanModel = Field(default_factory=TestPlanModel)
    user_defined_variables: List[UserDefinedVariable] = []
    http_defaults: Optional[HttpDefaultsModel] = None
    cookie_manager: Optional[CookieManagerModel] = None
    cache_manager: Optional[CacheManagerModel] = None
    csv_data_sets: List[CSVDataSetModel] = []
    thread_groups: List[ThreadGroupModel] = []
    config_elements: List[ConfigElementModel] = []   # otros que no caben en cookie/cache
    listeners: List[ListenerModel] = []
    unmapped: List[UnsupportedElement] = []          # elementos top-level no soportados
    metadata: StructureMetadata = Field(default_factory=StructureMetadata)

File: frontend/src/types/aiScriptStructure.ts
/**
 * Tipos TypeScript para AIScriptStructure.
 * Espejo de backend/app/schemas/ai_script_structure.py.
 *
 * El modelo NO se persiste en DB: el frontend lo solicita on-demand al backend
 * llamando a POST /script-designer/ai/parse-jmx (Sprint 2.2) y lo envía de vuelta
 * en POST /script-designer/ai/regenerate-jmx (Sprint 2.3).
 *
 * Sprint 2.3a: campo `is_dirty?: boolean` añadido en 24 interfaces para habilitar
 * el regenerador edit-preserving. Es opcional para no romper código del Sprint 2.0.
 */

// ============================================================================
// Aliases para legibilidad
// ============================================================================

export type UUID = string;

// ============================================================================
// Test Plan
// ============================================================================

export interface TestPlanModel {
  name: string;
  functional_mode: boolean;
  serialize_threadgroups: boolean;
  tearDown_on_shutdown: boolean;
  comments?: string | null;
  is_dirty?: boolean;
}

// ============================================================================
// Variables
// ============================================================================

export interface UserDefinedVariable {
  name: string;
  value: string;
  metadata: string;
  description?: string | null;
  is_dirty?: boolean;
}

// ============================================================================
// HTTP Defaults
// ============================================================================

export interface HttpDefaultsModel {
  domain?: string | null;
  protocol?: string | null;
  port?: string | null;
  path?: string | null;
  implementation?: string | null;
  encoding?: string | null;
  is_dirty?: boolean;
}

// ============================================================================
// Cookie / Cache Manager
// ============================================================================

export interface CookieManagerModel {
  enabled: boolean;
  clear_each_iteration: boolean;
  policy?: string | null;
  is_dirty?: boolean;
}

export interface CacheManagerModel {
  enabled: boolean;
  clear_each_iteration: boolean;
  use_expires: boolean;
  is_dirty?: boolean;
}

// ============================================================================
// CSV Data Sets
// ============================================================================

export type CSVShareMode = 'shareMode.all' | 'shareMode.group' | 'shareMode.thread';

export interface CSVDataSetModel {
  id: UUID;
  testname: string;
  enabled: boolean;
  filename: string;
  variable_names: string[];
  delimiter: string;
  ignore_first_line: boolean;
  quoted_data: boolean;
  recycle: boolean;
  share_mode: CSVShareMode;
  stop_thread: boolean;
  file_encoding?: string | null;
  is_dirty?: boolean;
}

// ============================================================================
// Stepping config — los aliases del backend se respetan al recibir JSON
// (Pydantic serializa con field name por defecto; revisaremos en Sprint 2.1).
// ============================================================================

export interface SteppingConfig {
  initial_delay: number;
  start_users_count: number;
  start_users_count_burst: number;
  start_users_period: number;
  stop_users_count: number;
  stop_users_period: number;
  ramp_up: number;
  flight_time: number;
}

// ============================================================================
// Sampler children
// ============================================================================

export interface HeaderModel {
  name: string;
  value: string;
}

export interface HeaderManagerModel {
  id: UUID;
  enabled: boolean;
  headers: HeaderModel[];
  is_dirty?: boolean;
}

export type AssertionTestField =
  | 'Assertion.response_data'
  | 'Assertion.response_code'
  | 'Assertion.response_headers'
  | 'Assertion.response_message';

export type AssertionPatternMatch = 'contains' | 'matches' | 'equals' | 'substring';

export interface ResponseAssertionModel {
  id: UUID;
  enabled: boolean;
  name: string;
  test_field: AssertionTestField | string;
  test_type: number;
  test_strings: string[];
  custom_message?: string | null;
  assume_success: boolean;
  negate: boolean;
  pattern_match: AssertionPatternMatch;
  is_dirty?: boolean;
}

export type ExtractFrom = 'body' | 'header' | 'url' | 'code' | 'message';
export type RegexUseHeaders = 'false' | 'true' | 'URL' | 'code' | 'message';

export interface RegexExtractorModel {
  id: UUID;
  enabled: boolean;
  name: string;
  refname: string;
  regex: string;
  template: string;
  match_number: string;
  default: string;
  default_empty_value: boolean;
  use_headers: RegexUseHeaders;
  scope?: 'all' | 'parent' | 'children' | null;
  extract_from: ExtractFrom;
  is_dirty?: boolean;
}

export interface JsonExtractorModel {
  id: UUID;
  enabled: boolean;
  name: string;
  refname: string;
  json_path: string;
  match_number: string;
  default: string;
  is_dirty?: boolean;
}

export interface XPathExtractorModel {
  id: UUID;
  enabled: boolean;
  name: string;
  refname: string;
  xpath: string;
  default: string;
  is_dirty?: boolean;
}

export interface BoundaryExtractorModel {
  id: UUID;
  enabled: boolean;
  name: string;
  refname: string;
  left_boundary: string;
  right_boundary: string;
  match_number: string;
  default: string;
  is_dirty?: boolean;
}

export interface ConstantTimerModel {
  id: UUID;
  enabled: boolean;
  name: string;
  delay_ms: number;
  is_dirty?: boolean;
}

export interface UniformRandomTimerModel {
  id: UUID;
  enabled: boolean;
  name: string;
  constant_delay_ms: number;
  random_delay_ms: number;
  is_dirty?: boolean;
}

export interface GaussianRandomTimerModel {
  id: UUID;
  enabled: boolean;
  name: string;
  constant_delay_ms: number;
  deviation_ms: number;
  is_dirty?: boolean;
}

// ============================================================================
// Sampler body
// ============================================================================

export interface FormArgument {
  name: string;
  value: string;
  always_encode: boolean;
  use_equals: boolean;
  metadata: string;
}

export type BodyMode = 'raw' | 'form' | 'none';
export type BodyType = 'json' | 'xml' | 'form' | 'raw' | 'auto';

export interface SamplerBody {
  mode: BodyMode;
  raw_text?: string | null;
  form_args: FormArgument[];
  body_type: BodyType;
}

// ============================================================================
// Sampler child (discriminado)
// ============================================================================

export type SamplerChildType =
  | 'header_manager'
  | 'response_assertion'
  | 'regex_extractor'
  | 'json_extractor'
  | 'xpath_extractor'
  | 'boundary_extractor'
  | 'constant_timer'
  | 'uniform_random_timer'
  | 'gaussian_random_timer'
  | 'unsupported';

export interface UnsupportedElement {
  id: UUID;
  type: 'unsupported';
  kind: string;
  name?: string | null;
  reason: string;
  severity: 'warning' | 'info';
  raw_xml: string;
  parent_sampler_id?: UUID | null;
}

export type SamplerChildData =
  | HeaderManagerModel
  | ResponseAssertionModel
  | RegexExtractorModel
  | JsonExtractorModel
  | XPathExtractorModel
  | BoundaryExtractorModel
  | ConstantTimerModel
  | UniformRandomTimerModel
  | GaussianRandomTimerModel
  | UnsupportedElement;

export interface SamplerChild {
  type: SamplerChildType;
  order: number;
  data: SamplerChildData;
}

// ============================================================================
// HTTP Sampler
// ============================================================================

export interface HTTPSamplerModel {
  id: UUID;
  type: 'sampler';
  enabled: boolean;
  name: string;
  method: string;
  protocol?: string | null;
  domain?: string | null;
  port?: string | null;
  path: string;
  content_encoding: string;
  follow_redirects: boolean;
  use_keepalive: boolean;
  auto_redirects: boolean;
  body: SamplerBody;
  children: SamplerChild[];
  raw_xml: string;
  is_dirty?: boolean;
}

// ============================================================================
// Controllers
// ============================================================================

export type ControllerKind =
  | 'generic_controller'
  | 'loop_controller'
  | 'if_controller'
  | 'while_controller'
  | 'throughput_controller';

interface ControllerBase {
  id: UUID;
  type: 'controller';
  enabled: boolean;
  name: string;
  children: TGChild[];
  is_dirty?: boolean;
}

export interface GenericControllerModel extends ControllerBase {
  kind: 'generic_controller';
}

export interface LoopControllerModel extends ControllerBase {
  kind: 'loop_controller';
  loops: number;
  continue_forever: boolean;
}

export interface IfControllerModel extends ControllerBase {
  kind: 'if_controller';
  condition: string;
  use_expression: boolean;
  evaluate_all: boolean;
}

export interface WhileControllerModel extends ControllerBase {
  kind: 'while_controller';
  condition: string;
}

export interface ThroughputControllerModel extends ControllerBase {
  kind: 'throughput_controller';
  style: number;
  percent_throughput: string;
  per_thread: boolean;
}

export type ControllerUnion =
  | GenericControllerModel
  | LoopControllerModel
  | IfControllerModel
  | WhileControllerModel
  | ThroughputControllerModel;

// ============================================================================
// Thread Group child (discriminado)
// ============================================================================

export interface TGChild {
  type: 'sampler' | 'controller' | 'unsupported';
  order: number;
  sampler?: HTTPSamplerModel | null;
  controller?: ControllerUnion | null;
  unsupported?: UnsupportedElement | null;
}

// ============================================================================
// Thread Group
// ============================================================================

export type ThreadGroupKind = 'standard' | 'stepping' | 'concurrency' | 'ultimate';

export type OnSampleError =
  | 'continue'
  | 'startnextloop'
  | 'stopthread'
  | 'stoptest'
  | 'stoptestnow';

export interface ThreadGroupModel {
  id: UUID;
  kind: ThreadGroupKind;
  name: string;
  enabled: boolean;
  comments?: string | null;
  on_sample_error: OnSampleError;
  num_threads: number;
  ramp_time: number;
  duration?: number | null;
  delay?: number | null;
  scheduler: boolean;
  loops: number;
  continue_forever: boolean;
  stepping?: SteppingConfig | null;
  children: TGChild[];
  raw_xml: string;
  is_dirty?: boolean;
}

// ============================================================================
// Config elements
// ============================================================================

export type ConfigElementKind =
  | 'cookie_manager'
  | 'cache_manager'
  | 'auth_manager'
  | 'dns_cache_manager'
  | 'keystore_config'
  | 'other';

export interface ConfigElementModel {
  id: UUID;
  kind: ConfigElementKind;
  name: string;
  enabled: boolean;
  properties: Record<string, any>;
  raw_xml: string;
  is_dirty?: boolean;
}

// ============================================================================
// Listeners
// ============================================================================

export type ListenerKind =
  | 'view_results_tree'
  | 'summary_report'
  | 'aggregate_report'
  | 'graph_results'
  | 'kg_apc_response_times_over_time'
  | 'kg_apc_response_codes_per_second'
  | 'kg_apc_transactions_per_second'
  | 'kg_apc_active_threads_over_time'
  | 'kg_apc_hits_per_second'
  | 'other';

export interface ListenerModel {
  id: UUID;
  kind: ListenerKind;
  guiclass: string;
  name: string;
  enabled: boolean;
  filename?: string | null;
  raw_xml: string;
  is_dirty?: boolean;
}

// ============================================================================
// Metadata derivada
// ============================================================================

export interface StructureMetadata {
  jmx_version: string;
  parsed_at: string;
  referenced_variables: string[];
  defined_variables: string[];
  undefined_variables: string[];
  has_unmapped: boolean;
  unmapped_count: number;
  parse_warnings: string[];
}

// ============================================================================
// Modelo raíz
// ============================================================================

export interface AIScriptStructure {
  test_plan: TestPlanModel;
  user_defined_variables: UserDefinedVariable[];
  http_defaults?: HttpDefaultsModel | null;
  cookie_manager?: CookieManagerModel | null;
  cache_manager?: CacheManagerModel | null;
  csv_data_sets: CSVDataSetModel[];
  thread_groups: ThreadGroupModel[];
  config_elements: ConfigElementModel[];
  listeners: ListenerModel[];
  unmapped: UnsupportedElement[];
  metadata: StructureMetadata;
}
